Unit test ValidatorRecurringEnrollment sometimes fails

hewison-chris opened the pull request:
This improves the reliability of the test where two nodes are put to sleep and then woken up during the test. To get a green build some other tests and timers and timeouts were also updated.

hewison-chris commented on source/agora/test/Base.d:
```diff
@@ -1071,7 +1071,7 @@ public class TestAPIManager
                 client_idxs.each!(idx =>
                     retryFor(this.clients[idx].hasTransactionHash(tx.hashFull()),
-                        4.seconds, format!"[%s:%s] Client #%s did not receive tx in expected time for height %s"
+                        2.seconds, format!"[%s:%s] Client #%s did not receive tx in expected time for height %s"
```
This is just temporary to make the test fail more often.

hewison-chris commented on source/agora/test/ValidatorRecurringEnrollment.d:
```diff
@@ -234,44 +234,47 @@ unittest
     auto enrolls1 = blocks[$ - 1].header.enrollments.length;
 
-    // This nodes will wake up to an expired cycle, it should immediately enroll
+    assert(enrolls1 == GenesisValidators - 1);
```
This is not as expected. All validators are enrolled!

hewison-chris replied:
Thanks @omerfirmak your latest PR fixed this.

hewison-chris replied:
I spoke too soon. It can still fail on this check. So it seems that the validator can re-enroll even when sleeping.

Geod24 replied:
Yeah, as discussed yesterday, `sleep` only makes the server-side sleep... It does nothing to prevent the timers from firing.

hewison-chris replied:
Updated to skip `checkAndEnroll` in this test during the time the nodes are put to sleep.

Geod24 replied:
That's solved AFAIK.

hewison-chris commented on source/agora/node/Validator.d:
```diff
@@ -605,12 +605,16 @@ public class Validator : FullNode, API
-        this.enroll_man.enroll_pool.addValidated(enrollment, avail_height);
-        this.network.peers.each!(p => p.client.sendEnrollment(enrollment));
-        return enrollment;
+        if (height + 2 >= avail_height) // We are near the end of the cycle
```
I added this as we were enrolling at every block height.

Geod24 replied:
Use early return

hewison-chris committed:
Only auto enroll near the end of the cycle

When a validator is configured to auto enroll it should send enrollments
when near the end of the cycle rather then every block.
Also the `ValidatorRecurringEnrollment` test was not really testing
enrolling when coming back on line as the nodes were enrolling before
they were put in sleep mode.

hewison-chris committed:
Unit test update: ValidatorRecurringEnrollment

When node #4 is catching up we don't need to check the first 4 nodes.

hewison-chris committed:
Unit test update: ValidatorRecurringEnrollment

Updated node #5 catch up to use `expectHeightAndPreImg` rather than
`expectHeight` which is more appropriate as it makes sure `pre-images`
are available before `this.setTimeFor(height)` is called.

kchulj commented on source/agora/test/ValidatorRecurringEnrollment.d:
```diff
@@ -238,16 +281,16 @@ unittest
 
     // Approach end of the cycle
-    network.generateBlocks(Height(GenesisValidatorCycle - 2));
+    network.generateBlocks(Height(GenesisValidatorCycle - 3));
```
Why was this changed to 3?

hewison-chris replied:
The nodes that are auto enrolling will send enrollments from 2 block heights before the end of the cycle. Although we actually prevent the node from sending enrollments before the height we wake the node up it makes more sense to only generate blocks to Height 17 before putting them to sleep.

linked0 commented on source/agora/test/ValidatorRecurringEnrollment.d:
```diff
@@ -234,44 +277,46 @@ unittest
         Height(GenesisValidatorCycle - 1), network.blocks[0].header);
 
-    network.generateBlocks(iota(GenesisValidators - 1),
+    network.generateBlocks(iota(GenesisValidators - 1), // nodes #0 -> #4
```
```suggestion
    network.generateBlocks(iota(GenesisValidators - 1), // nodes #0 .. #4
```
At the first sight, I'm thinking what the meaning of `->` is. But I would suggest removing this kind of comment because if we make change the line, we could forget to modify the comment.

hewison-chris replied:
Updated

linked0 commented on source/agora/test/ValidatorRecurringEnrollment.d:
```diff
@@ -234,44 +277,46 @@ unittest
 
-    auto sleep_node_1 = nodes[$ - 1];
-    auto sleep_node_2 = nodes[$ - 2];
+    auto sleep_node_1 = nodes[$ - 1]; // node #5
+    auto sleep_node_2 = nodes[$ - 2]; // node #4
```
```suggestion
    auto sleep_node_1 = nodes[$ - 2]; // node #4
    auto sleep_node_2 = nodes[$ - 1]; // node #5
```
I think this would be more intuitive. But do we need the comment like "node #5, node #4"?
I know that the suggested change makes you change all the lines of this test. I only suggest that if you have extra time.

hewison-chris replied:
In fact I wanted to rename all the node names in this test as it is very confusing but didn't want too many changes at one time. Now I think I can add a final commit to do just that.

hewison-chris replied:
I have added a couple of rename commits to make it clearer.

linked0 reviewed (approved):
Great! This is what I have been waiting for. Ty.

hewison-chris committed:
Trivial rename of node_1 to node_0

hewison-chris committed:
Trivial rename of some nodes in test

hewison-chris commented:
Now this one fails for macOS ldc-master in a new place:-
`core.exception.AssertError@source/agora/test/ValidatorRecurringEnrollment.d(296): Check condition failed after timeout of 10 secs and 100 attempts: Node 4 has block height 18. Expected: 19`

hewison-chris commented:
This will be updated to use `suspend` in `LocalRest` when implemented.

hewison-chris commented:
I see in https://github.com/Geod24/localrest/pull/127 the `sleep` has been updated to also suspend all tasks and timers.
This PR should be updated to use that version of `localrest`.

hewison-chris commented:
Updated to use new `localrest` with `sleep` also stopping tasks and timers.

Geod24 commented:
Looks like it still fails ?

hewison-chris commented:
```
2021-08-17 04:45:32,939 Error [agora.consensus.protocol.Nominator] - validateValue(): Validation failed: May be valid. Data: { tx_set: [0x3b1e...db7b], enrolls: [], missing_validators: [], time_offset: 10200 }
2021-08-17 04:45:39,474 Error [agora.consensus.protocol.Nominator] - validateValue(): Validation failed: May be valid. Data: { tx_set: [0x3b1e...db7b], enrolls: [], missing_validators: [], time_offset: 10200 }
2021-08-17 04:45:39,475 Info [agora.consensus.protocol.Nominator] - Nominating { tx_set: [], enrolls: [], missing_validators: [], time_offset: 12000 } at 1609471200
2021-08-17 04:45:39,475 Info [agora.consensus.protocol.Nominator] - agora.consensus.protocol.Nominator.Nominator.nominate(): Proposing tx set for slot 18
2021-08-17 04:45:39,475 Info [agora.consensus.protocol.Nominator] - agora.consensus.protocol.Nominator.Nominator.nominate(): Tx set didn't trigger new nomination
2021-08-17 04:45:39,476 Info [agora.network.Manager] - Doing network discovery..
2021-08-17 04:45:39,488 Info [agora.network.Manager] - Retrieving blocks [18..20] from Validator #0 (boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh)..
2021-08-17 04:45:39,488 Info [agora.network.Manager] - Retrieving preimages from the height of 0 from Validator #0 (boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh)..
2021-08-17 04:45:39,488 Info [agora.network.Manager] - Received 6 preimages
2021-08-17 04:45:39,497 Info [agora.network.Manager] - Received blocks [18..20]
2021-08-17 04:45:39,500 Info [agora.consensus.Ledger] - Beginning externalization of block #18
2021-08-17 04:45:39,500 Info [agora.consensus.Ledger] - Transactions: 1 - Enrollments: 0
2021-08-17 04:45:39,500 Info [agora.consensus.Ledger] - Validators: Active: 6 - Signing: 111110 - Slashed: []
2021-08-17 04:45:39,504 Info [agora.consensus.Ledger] - Beginning externalization of block #19
2021-08-17 04:45:39,504 Info [agora.consensus.Ledger] - Transactions: 1 - Enrollments: 0
2021-08-17 04:45:39,504 Info [agora.consensus.Ledger] - Validators: Active: 6 - Signing: 111110 - Slashed: []
2021-08-17 04:45:39,511 Info [agora.consensus.Ledger] - Beginning externalization of block #20
2021-08-17 04:45:39,511 Info [agora.consensus.Ledger] - Transactions: 1 - Enrollments: 5
2021-08-17 04:45:39,511 Info [agora.consensus.Ledger] - Validators: Active: 0 - Signing: 111110 - Slashed: []
======================================================================

Module tests failed: agora.test.ValidatorRecurringEnrollment
core.exception.AssertError@source/agora/test/Base.d(1073): Check condition failed after timeout of 4 secs and 40 attempts: [source/agora/test/ValidatorRecurringEnrollment.d:269] Client #5 did not receive tx in expected time for height 21
```

hewison-chris committed:
`NetworkManager` test: Increase frequency of block catchup

As this test is relying on the block catchup then a shorter interval
will help ensure the catchup occurs in time.

hewison-chris committed:
`NetworkClientTest` wait for preimages first

Use `network.expectHeightAndPreImg` which will ensure preimages are
available before setting nomination time.
Also increase interval of block catchup as the test is not relying on
block updates via that route.

hewison-chris committed:
Decrease time between retries in tests

This is to help decrease the time needed for the unit tests to run and
also to attempt to make them more reliable.

Geod24 commented on source/agora/node/Validator.d:
```diff
@@ -605,12 +605,16 @@ public class Validator : FullNode, API
+            this.network.peers.each!(p => p.client.sendEnrollment(enrollment));
+            return enrollment;
+        }
+        return Enrollment.init;
```
Ideally this function shouldn't return anything, but I think something in the test uses it.

Geod24 replied:
This is just a general message for later though.

hewison-chris replied:
Yes some tests use the returned enrollment to check other nodes have it in their pools.

Geod24 commented on source/agora/utils/Utility.d:
```diff
@@ -131,8 +131,8 @@ public void retryFor (Exc : Throwable = AssertError) (lazy bool check,
 {
-    // wait 100 msecs between attempts
-    const SleepTime = 100;
+    // wait between attempts
+    const SleepTime = 50;
     auto attempts = timeout.total!"msecs" / SleepTime;
```
This commit has a lot of impact, so it should come with a descriptive message.

Geod24 reviewed (approved):
LGTM. Some commit could use a better message. E.g. "NetworkClientTest updates" doesn't really tell much.

hewison-chris commented:
Looks like this test was passing for the wrong reason before:- 
`core.exception.AssertError@source/agora/test/Base.d(1156): Check condition failed after timeout of 5 secs and 100 attempts: [source/agora/test/Fee.d:135] Expected height 0 for client #0 not 1`

hewison-chris committed:
testConf: pre-image reveal interval

In the network tests it would be useful to have the frequency of checks
to reveal `pre-images` configurable.

hewison-chris committed:
`Fee` test: Disable for mismatching configuration

The second unit test is testing that a block is not externalized for a
node with a different configuration for fee payouts. This is not
currently implemented and the test is failing now some test
configuration has been updated and the block is created faster so within
the 1 second sleep.

hewison-chris committed:
Test conf: nomination interval

This adds to `TestConf` `nomination_interval` so that tests can
configure this frequency as required.

hewison-chris committed:
Test conf: reduce `preimage_catchup_interval`

The interval between checking for missing pre-images during the
nomination phase can be greatly reduced for network tests.

hewison-chris committed:
Test conf: decrease block catchup interval

When nodes are missing transactions locally they catchup during the
periodic block catchup. Setting this value lower for network tests will
improve the speed of catchup and help some tests from timing out.
For the `LocalTransactions` test which is specifically testing 6 nodes
having different transactions I have set the value even lower at 100 ms.

hewison-chris commented:
Finally all required have passed (on first attempt). I think this will help us have less failures going forward.
